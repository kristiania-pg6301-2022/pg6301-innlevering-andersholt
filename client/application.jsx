import * as React from "react";
import { Link } from "react-router-dom";
import { useEffect, useState } from "react";

export function FrontPage() {
  return (
    <div>
      <h1>Kristiania Quiz</h1>
      <div>
        <Link to={"/question"}>Answer a question</Link>
      </div>
      <div>
        <Link to={"/score"}>Show Score</Link>
      </div>
    </div>
  );
}

async function fetchJSON(url) {
  const res = await fetch(url);
  return await res.json();
}

export function Question() {
  const [question, setQuestion] = useState();

  useEffect(async () => {
    setQuestion(await fetchJSON("api/question"));
  }, []);

  if (!question) {
    return <div>Loading....</div>;
  }

  async function handleAnswer(answer) {
    const { id } = question;
    await fetch("api/question", {
      method: "post",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ id, answer }),
    });
    await setQuestion(await fetchJSON("api/question"));
  }

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((a) => (
          <div key={a}>
            <button onClick={() => handleAnswer(a)}>
              {question.answers[a]}
            </button>
          </div>
        ))}
      <Link to={"/score"}>Show current score</Link>
    </div>
  );
}

/*function Q() {
  const [question, setQuestion] = useState();

  async function handle() {
    const res = await fetch("/api/question");
    setQuestion(await res.json());
  }

  if (!question) {
    return <div>Loading....</div>;
  }

  return <Question question={question} />;
}*/

export function ShowScore() {
  const [cookie, setCookie] = useState({});

  useEffect(async () => {
    const res = await fetch("/api/score");
    setCookie(await res.json());
  }, []);

  const { answered, correct } = cookie;

  return (
    <div>
      <h1>
        You have: {correct} correct, out of: {answered} answers
      </h1>
      <div>
        <Link to={"/question"}>Answer another question</Link>
      </div>
      <Link to={"/"}>Go Back to frontpage</Link>
    </div>
  );
}
