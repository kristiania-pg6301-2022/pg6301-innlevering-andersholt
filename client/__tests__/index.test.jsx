import React from "react";
import ReactDOM, { render } from "react-dom";
import { MemoryRouter } from "react-router-dom";
import express from "express";
import bodyParser from "body-parser";
import cookieParser from "cookie-parser";
import request from "supertest";
import "babel-polyfill";

import { FrontPage, Question } from "../application.jsx";

const app = express();
app.use(bodyParser.json);
app.use(cookieParser("test secret"));
app.use("/question", Question);

describe("Quiz game Innlevering", () => {
  it("displays homepage", () => {
    const element = document.createElement("div");

    ReactDOM.render(
      <MemoryRouter>
        <FrontPage />
      </MemoryRouter>,
      element
    );

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("returns a random question", async () => {
    const response = await request(app).get("/question/random").expect(200);
    expect(response.body).toMatchObject({
      id: expect.any(Number),
      answers: expect.any(Object),
      category: expect.any(String),
    });
    expect(response.body).not.toHaveProperty("correct_answers");
  });
});
