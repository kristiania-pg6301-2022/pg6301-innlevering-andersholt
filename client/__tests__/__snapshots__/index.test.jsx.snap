// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index displays homepage 1`] = `"<div><h1>Kristiania Quiz</h1><div><a href=\\"/question\\">Answer a question</a></div><div><a href=\\"/score\\">Show Score</a></div></div>"`;
